submission_offsets.py

```diff
@@ -0,0 +1,68 @@
+from flask.cli import FlaskGroup
+import click
+from collections import defaultdict
+
+import db
+import webserver
+
+from sqlalchemy import text
+
+cli = FlaskGroup(add_default_commands=False, create_app=webserver.create_app_flaskgroup)
+
+@cli.command(name='add-offsets')
+@click.option("--limit", "-l", default=10000)
+def add_offsets(limit):
+	"""Update lowlevel submission offsets with a specified limit."""
+	incremental_add_offset(limit)
+
+def incremental_add_offset(limit):
+	with db.engine.connect() as connection:
+		# Find the next batch of items to update
+		batch_query = text("""
+			SELECT id, gid
+			  FROM lowlevel
+			 WHERE submission_offset IS NULL
+			 LIMIT :limit
+		""")
+		batch_result = connection.execute(batch_query, { "limit": limit })
+
+		# Find max existing offsets
+		offset_query = text("""
+			SELECT gid, MAX(submission_offset)
+			  FROM lowlevel
+			 WHERE submission_offset IS NOT NULL
+		      GROUP BY gid
+		""")
+		offset_result = connection.execute(offset_query)
+
+		max_offsets = defaultdict(int)
+		for gid, max_offset in offset_result.fetchall():
+			max_offsets[gid] = max_offset
+
+		print("Starting batch insertions...")
+		print("============================")
+		count = 0
+		for id, gid in batch_result.fetchall():
+			print("Finished {}/{} items".format(count, limit))
```

Comment:
will these print messages occur for every item in the 10000 batch?

Reply:
Yes, the printing occurs for each item in the 10000 batch as the offset is being found and inserted. If we are running this continuously with the added loop that you mentioned, would it be best for me to remove some of the printing? I.e. print for each batch starting and ending rather than on each item in a batch?

Reply:
exactly, there should be one print statement for each 10000 items